### PR TITLE
[Core] Update memory pool

### LINKED ISSUE #9929: Deprecation warnings in core due to LockObject's move constructor
**Description**
There is one last compilation warning in the Kratos Core. It is raised in two diferent places, both because they use the move constructor of LockObject, which is marked deprecated:
https://github.com/KratosMultiphysics/Kratos/blob/629db7132a30f900048e0bc14e4f5b17e33d4456/kratos/includes/lock_object.h#L58-L68

First warning (line 63):

https://github.com/KratosMultiphysics/Kratos/blob/629db7132a30f900048e0bc14e4f5b17e33d4456/kratos/includes/chunk.h#L62-L70

Second warning:

https://github.com/KratosMultiphysics/Kratos/blob/629db7132a30f900048e0bc14e4f5b17e33d4456/kratos/includes/thread_fixed_size_memory_pool.h#L65

This is deep within Kratos and I don't really know what these do, so I'll refrain from touching anything myself.

**Scope**
- Compilation
- KratosCore

**To Reproduce**
Compile from scratch

**Expected behavior**
No warning.

**Environment**
- OS: Ubuntu 22.04 LTS x86_64
- Branch: master
- Compiler: Both gcc and clang

### PR DESCRIPTION
**Pull request by pooyan-dadvand**
**📝 Description**
This PR refactors the memory pool to be thread friendly.  (Before the stress test for allocate/deallocate was failing) It also implies low cost spin mutex instead of OS ones. 

The performance depends highly on the type of usage:

- For large pools (like the one for Nodes) is slightly better than the standard new and delete pools but not justifying its usage in production now. (It may become better moving to c++17)

- For small pools with many new and deletes (like local variable allocations) it works 3-7 times faster. 

- In a nutshell, it would be better to use a stack pool rather than this one for local variables and keep large number of entities in the normal heap.

- Main reason is the fact that we are not add and delete nodes or elements regularly

So I have not substitute the standard new delete and leave the code to see if we can reuse it in future using C++17 features.

Meanwhile, it fixes #9929 

Please mark the PR with appropriate tags: 
- Api Breaker, Mpi, etc...

**🆕 Changelog**
Refactoring the memory pool
Removing the lock object move constructor

### CODE REVIEW

File: kratos/includes/chunk.h [hunk 1]
```diff
@@ -15,6 +15,8 @@
 #if !defined(KRATOS_CHUNK_H_INCLUDED )
 #define  KRATOS_CHUNK_H_INCLUDED
 
+#include <atomic>
```
Comment (loumalouomega) [inline kratos/includes/chunk.h:19]:
```suggestion
// System includes
#include <atomic>
```

File: kratos/includes/chunk.h [hunk 2]
```diff
@@ -15,6 +15,8 @@
...
+
 #include "includes/lock_object.h"
```
Comment (loumalouomega) [inline kratos/includes/chunk.h:24]:
```suggestion
// External includes

// Project includes
#include "includes/lock_object.h"
```

File: kratos/includes/chunk.h [hunk 3]
```diff
@@ -59,24 +61,18 @@ namespace Kratos
...
+		  : mpData(nullptr)
+		  , mpEnd(nullptr)
+		  , mpUninitializedMemory(nullptr)
 		  , mSize(SizeInBytes)
-		  , mBlockSizeInBytes(BlockSizeInBytes){
-		  mOwnerThread = OpenMPUtils::ThisThread();
-	  }
+		  , mBlockSizeInBytes(BlockSizeInBytes)
+		  , mNumberOfAvailableBlocks(0) // to be initialized in initialize
+		  , mFirstAvailableBlock(mpData)
```
Comment (loumalouomega) [inline kratos/includes/chunk.h:78]: There are tabs mixed with spaces

File: kratos/includes/chunk.h [hunk 4]
```diff
@@ -94,68 +90,64 @@ namespace Kratos
...
-		  for (auto i_thread = 0; i_thread < OpenMPUtils::GetCurrentNumberOfThreads(); i_thread++)
-			  SetNumberOfAvailableBlocks(0, i_thread);
-		  SetNumberOfAvailableBlocks(AllocatableDataSize() / block_size_after_alignment);
-
-		  SizeType i_block = 0;
-		  for (auto p = GetData(); i_block < GetNumberOfBlocks(); p += block_size_after_alignment)
-			  *p = ++i_block;
-
-		  KRATOS_DEBUG_CHECK_EQUAL(i_block, GetNumberOfAvailableBlocks());
+		  const std::size_t data_size = DataSize();
```
Comment (loumalouomega) [inline kratos/includes/chunk.h:97]: Idem, tabs mixed with spaces

File: kratos/includes/chunk.h [hunk 5]
```diff
@@ -94,68 +90,64 @@ namespace Kratos
...
-					  break;
-				  }
-			  unlock();
-		  }
-		  DataType * p_result = GetData() + (GetFirstAvailableBlockIndex() * GetBlockSize(mBlockSizeInBytes));
-		  KRATOS_DEBUG_CHECK(Has(p_result));
-		  SetFirstAvailableBlockIndex(*p_result);
-		  SetNumberOfAvailableBlocks(GetNumberOfAvailableBlocks()-1);
-
+			lock();
```
Comment (loumalouomega) [inline kratos/includes/chunk.h:115]: Again, tabs+spaces

File: kratos/includes/chunk.h [hunk 6]
```diff
@@ -94,68 +90,64 @@ namespace Kratos
...
-		  SetFirstAvailableBlockIndex(*p_result);
-		  SetNumberOfAvailableBlocks(GetNumberOfAvailableBlocks()-1);
-
+			lock();  
+			BlockType * p_result = mFirstAvailableBlock;
+			mFirstAvailableBlock = (p_result >= mpUninitializedMemory) ? mFirstAvailableBlock + mBlockSizeAfterAlignment : (BlockType*)*p_result;
+			// if(p_result >= mpUninitializedMemory) {
+			// 	mFirstAvailableBlock += mBlockSizeAfterAlignment;
+			// 	if(mpUninitializedMemory < mpEnd) {
+			// 		mpUninitializedMemory += mBlockSizeAfterAlignment;
```
Comment (loumalouomega) [inline kratos/includes/chunk.h:121]: Remove commented code?

File: kratos/includes/chunk.h [hunk 7]
```diff
@@ -94,68 +90,64 @@ namespace Kratos
...
+	
 		  return p_result;
 	  }
 
 	  void Deallocate(void* pPointrerToRelease) {
 		  if (pPointrerToRelease == nullptr)
 			  return;
 
-		  KRATOS_DEBUG_CHECK_NOT_EQUAL(GetData(), nullptr);
+		  KRATOS_DEBUG_CHECK_NOT_EQUAL(mpData, nullptr);
```
Comment (loumalouomega) [inline kratos/includes/chunk.h:142]: More tabs+spaces

File: kratos/includes/chunk.h [hunk 8]
```diff
@@ -273,33 +269,26 @@ namespace Kratos
       ///@name Member Variables
       ///@{
 
-		DataType * mpData;
+		BlockType * mpData;
```
Comment (loumalouomega) [inline kratos/includes/chunk.h:276]: More tabs+spaces

File: kratos/includes/thread_fixed_size_memory_pool.h [hunk 9]
```diff
@@ -92,75 +96,114 @@ namespace Kratos
...
+
+		if(mpCurrentChunk == nullptr){
+			AddChunk();
+		}
+
+		if (mpCurrentChunk->IsFull())
+			if (!mAvailableChunks.try_dequeue(mpCurrentChunk))
+				AddChunk();
+
+		// lock();
```
Comment (loumalouomega) [inline kratos/includes/thread_fixed_size_memory_pool.h]: Remove unused?

File: kratos/includes/thread_fixed_size_memory_pool.h [hunk 10]
```diff
@@ -92,75 +96,114 @@ namespace Kratos
...
+
+		if (mpCurrentChunk->IsReleased())
+			mpCurrentChunk->Initialize();
+		KRATOS_CHECK_IS_FALSE(mpCurrentChunk->IsFull());
+		p_result = mpCurrentChunk->Allocate();
+		KRATOS_DEBUG_CHECK_NOT_EQUAL(p_result, nullptr);
+
+		// unlock();
```
Comment (loumalouomega) [inline kratos/includes/thread_fixed_size_memory_pool.h]: More unused and commented

File: kratos/includes/thread_fixed_size_memory_pool.h [hunk 11]
```diff
@@ -92,75 +96,114 @@ namespace Kratos
...
-			  if ((*i_chunk)->Has(pPointrerToRelease)) {
-				  DeallocateFromAvailableChunk(pPointrerToRelease, i_chunk);
+	  bool Deallocate(void* pPointerToRelease) {
+		  if(mAvailablePointers.size_approx() < MaximumPointersToKeep){
+			  mAvailablePointers.enqueue(pPointerToRelease);
+			  return true;
+		  }
+
+		//   lock();
+		//   if (!mAvailableChunks.empty()) {
```
Comment (loumalouomega) [inline kratos/includes/thread_fixed_size_memory_pool.h]: Leftover?

File: kratos/includes/thread_fixed_size_memory_pool.h [hunk 12]
```diff
@@ -92,75 +96,114 @@ namespace Kratos
...
+				//   unlock();
 				  return true;
 			  }
-		  }
 
 		  for (auto i_chunk = mChunks.begin(); i_chunk != mChunks.end(); i_chunk++)
 		  {
-			  if (i_chunk->Has(pPointrerToRelease)) {
+			  if (i_chunk->Has(pPointerToRelease)) {
+				//   std::cout << "deleting pointer " << pPointerToRelease << " from " << *i_chunk << std::endl;
```
Comment (loumalouomega) [inline kratos/includes/thread_fixed_size_memory_pool.h]: Leftover?
```suggestion
```

File: kratos/includes/thread_fixed_size_memory_pool.h [hunk 13]
```diff
@@ -92,75 +96,114 @@ namespace Kratos
...
 
 	  void Release() {
 		  mChunks.clear();
 		  mNumberOfReleasedChunks += mChunks.size();
-		  mAvailableChunks.clear();
-		  for (auto i_chunk = mChunks.begin(); i_chunk != mChunks.end(); i_chunk++)
-			  mAvailableChunks.push_front(&(*i_chunk));
+		  Chunk* p_dummy;
+		  while(mAvailableChunks.try_dequeue(p_dummy));
+		//   for (auto i_chunk = mChunks.begin(); i_chunk != mChunks.end(); i_chunk++)
```
Comment (loumalouomega) [inline kratos/includes/thread_fixed_size_memory_pool.h]: Leftover?

File: kratos/includes/chunk.h (hunk 8 above)
Comment (loumalouomega) [reply kratos/includes/chunk.h:276]: In general in this PR there is a lot of these in this PR

File: kratos/python_scripts/run_cpp_tests.py [hunk 14]
```diff
@@ -1,3 +1,9 @@
-from KratosMultiphysics import *
-Tester.SetVerbosity(Tester.Verbosity.TESTS_OUTPUTS)
-Tester.RunAllTestCases()
+import KratosMultiphysics as Kratos
```
Comment (loumalouomega) [inline kratos/python_scripts/run_cpp_tests.py:1]: This could be an independent PR

**Review by loumalouomega — COMMENTED** [review]
Some comments

**Comment by loumalouomega** [comment]
> Some comments

I do not block as I think this should be blocked by @KratosMultiphysics/technical-committee in any case

**Comment by pooyan-dadvand** [comment]
Pushing commits with the tests

File: kratos/includes/chunk.h (hunk 6 above)
Comment (pooyan-dadvand) [reply kratos/includes/chunk.h:121]: I leaved it as it explains the logic of the above line. (with a small change which I leave to test)

File: kratos/python_scripts/run_cpp_tests.py (hunk 14 above)
Comment (pooyan-dadvand) [reply kratos/python_scripts/run_cpp_tests.py:1]: Yes, but it is useful to run stress tests and check the time

File: kratos/includes/chunk.h (hunk 8 above)
Comment (pooyan-dadvand) [reply kratos/includes/chunk.h:276]: I can change them to space but almost all the lines got changed which makes the reviewing more difficult

File: kratos/includes/thread_fixed_size_memory_pool.h (hunk 9 above)
Comment (pooyan-dadvand) [reply kratos/includes/thread_fixed_size_memory_pool.h]: Removed

File: kratos/includes/thread_fixed_size_memory_pool.h (hunk 10 above)
Comment (pooyan-dadvand) [reply kratos/includes/thread_fixed_size_memory_pool.h]: Removed

File: kratos/includes/thread_fixed_size_memory_pool.h (hunk 12 above)
Comment (pooyan-dadvand) [reply kratos/includes/thread_fixed_size_memory_pool.h]: Just removed in the commit with others

File: kratos/includes/thread_fixed_size_memory_pool.h (hunk 13 above)
Comment (pooyan-dadvand) [reply kratos/includes/thread_fixed_size_memory_pool.h]: Removed!

**Comment by roigcarlo** [comment]
@pooyan-dadvand Centos in now fixed in here as well. If we can merge I will add this to https://github.com/KratosMultiphysics/Kratos/pull/9937

**Review by roigcarlo — APPROVED** [review]
:+1:

**Pull request merged**